theano/sandbox/rng_mrg.py

```diff
@@ -1446,55 +1446,85 @@ def multinomial(self, size=None, n=1, pvals=None, ndim=None, dtype='int64',
         check that the elements are non-negative, less than 1, or
         sum to 1. passing pvals = [[-2., 2.]] will result in
         sampling [[0, 0]]
 
+        - When `a` and `p` are tensors their shape should be the same.
+
+        - Only replace=False is implemented for now.
+
         """
-        if pvals is None:
-            raise TypeError("You have to specify pvals")
-        pvals = as_tensor_variable(pvals)
+        if replace:
+            raise NotImplementedError(
+                "MRG_RandomStreams.choice only works without replacement "
+                "for now.")
+
+        if p is None:
+            raise TypeError("You have to specify p.")
+        p = as_tensor_variable(p)
 
-        if size is not None:
-            raise ValueError("Provided a size argument to "
-                             "MRG_RandomStreams.multinomial_wo_replacement, "
-                             "which does not use the size argument.")
         if ndim is not None:
-            raise ValueError("Provided an ndim argument to "
-                             "MRG_RandomStreams.multinomial_wo_replacement, "
-                             "which does not use the ndim argument.")
-        if pvals.ndim == 2:
-            # size = [pvals.shape[0], as_tensor_variable(n)]
-            size = pvals[:, 0].shape * n
-            unis = self.uniform(size=size, ndim=1, nstreams=nstreams)
-            op = multinomial.MultinomialWOReplacementFromUniform(dtype)
-            n_samples = as_tensor_variable(n)
-            return op(pvals, unis, n_samples)
-        else:
+            raise ValueError("ndim argument to "
+                             "MRG_RandomStreams.multinomial_wo_replacement "
+                             "is not used.")
+
+        if p.ndim == 1:
+            p = tensor.shape_padleft(p)
+
+        if p.ndim != 2:
             raise NotImplementedError(
                 "MRG_RandomStreams.multinomial_wo_replacement only implemented"
-                " for pvals.ndim = 2")
+                " for p.ndim = 1 or p.ndim = 2")
+
+        shape = p[:, 0].shape * size
+        unis = self.uniform(size=shape, ndim=1, nstreams=nstreams)
+        op = multinomial.MultinomialWOReplacementFromUniform(dtype)
+        sampled_indices = op(p, unis, as_tensor_variable(size))
+
+        if isinstance(a, int):
+            return sampled_indices
+        a = tensor.as_tensor_variable(a)
+        return a[sampled_indices]
+
+    def multinomial_wo_replacement(self, size=None, n=1, pvals=None,
+                                   ndim=None, dtype='int64', nstreams=None):
+        warnings.warn('MRG_RandomStreams.multinomial_wo_replacement() is '
+                      'deprecated and will be removed in the next release of '
+                      'Theano. Please use MRG_RandomStreams.choice() instead.')
+        return self.choice(size=n, replace=False, p=pvals, dtype=dtype,
+                           nstreams=nstreams, ndim=ndim)
```

Comment:
I'm not sure this conversion is well done. A way to test this is to make a test, that will create 2 MRG_RandomStreams() and draw from both of them one with choice and the other with multinomial_wo_replacements. And compare if we have the same output.

The code bellow is what I think is the equivalence, but compared with the numpy rng theano wrapper.

```
import theano
import theano.sandbox.rng_mrg
i=theano.tensor.iscalar()
mat=theano.tensor.matrix()
m=theano.sandbox.rng_mrg.MRG_RandomStreams()
mm=m.multinomial_wo_replacement(pvals=mat, n=i)
f=theano.function([mat,i], mm)

cm=theano.tensor.shared_randomstreams.RandomStreams()
cmm=cm.choice(a=mat.shape[1], size=[i], p=mat[0], replace=False)   # Here choice only support a vector for p, this limit the testing, but still allow enough testing I think.
cf=theano.function([mat,i], cmm)
p=[[.1]*10]
pp=[[0.01]*9+[.91,]]
for i in range(10): f(p, 2); cf(p,2)

print()
for i in range(10): f(pp, 4); cf(pp,4)
```